src/lib.rs

```diff
@@ -590,433 +598,34 @@ impl<'r> ::rocket::response::Responder<'r> for HttpApiProblem {
-        }
-    }
-}
-
-impl From<u16> for HttpStatusCode {
-    fn from(n: u16) -> HttpStatusCode {
-        use HttpStatusCode::*;
-        match n {
-            100 => Continue,
-            101 => SwitchingProtocols,
-            102 => Processing,
-            200 => Ok,
-            201 => Created,
-            202 => Accepted,
-            203 => NonAuthoritativeInformation,
-            204 => NoContent,
-            205 => ResetContent,
-            206 => PartialContent,
-            207 => MultiStatus,
-            208 => AlreadyReported,
-            226 => ImUsed,
-            300 => MultipleChoices,
-            301 => MovedPermanently,
-            302 => Found,
-            303 => SeeOther,
-            304 => NotModified,
-            305 => UseProxy,
-            307 => TemporaryRedirect,
-            308 => PermanentRedirect,
-            400 => BadRequest,
-            401 => Unauthorized,
-            402 => PaymentRequired,
-            403 => Forbidden,
-            404 => NotFound,
-            405 => MethodNotAllowed,
-            406 => NotAcceptable,
-            407 => ProxyAuthenticationRequired,
-            408 => RequestTimeout,
-            409 => Conflict,
-            410 => Gone,
-            411 => LengthRequired,
-            412 => PreconditionFailed,
-            413 => PayloadTooLarge,
-            414 => UriTooLong,
-            415 => UnsupportedMediaType,
-            416 => RangeNotSatisfiable,
-            417 => ExpectationFailed,
-            418 => ImATeapot,
-            421 => MisdirectedRequest,
-            422 => UnprocessableEntity,
-            423 => Locked,
-            424 => FailedDependency,
-            426 => UpgradeRequired,
-            428 => PreconditionRequired,
-            429 => TooManyRequests,
-            431 => RequestHeaderFieldsTooLarge,
-            451 => UnavailableForLegalReasons,
-            500 => InternalServerError,
-            501 => NotImplemented,
-            502 => BadGateway,
-            503 => ServiceUnavailable,
-            504 => GatewayTimeout,
-            505 => HttpVersionNotSupported,
-            506 => VariantAlsoNegotiates,
-            507 => InsufficientStorage,
-            508 => LoopDetected,
-            510 => NotExtended,
-            511 => NetworkAuthenticationRequired,
-            _ => Unregistered(n),
-        }
-    }
-}
-
-impl fmt::Display for HttpStatusCode {
-    fn fmt(&self, f: &mut fmt::Formatter) -> Result<(), fmt::Error> {
-        write!(f, "{} {}", self.to_u16(), self.title())
-    }
-}
-
-impl Serialize for HttpStatusCode {
-    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
+    pub fn serialize<S>(date: &Option<StatusCode>, s: S) -> Result<S::Ok, S::Error>
     where
         S: Serializer,
     {
-        serializer.serialize_u16(self.to_u16())
+        if let Some(ref status_code) = *date {
+            return s.serialize_u16(status_code.as_u16());
+        }
+        s.serialize_none()
```

Comment:
Not sure if we care but this is inconsistent with the `status()` method above. That one falls back to InternalServerError if it is not present whereas here, we just don't serialize.

Reply:
@chridou Why is `status` even an `Option` in the first place? Couldn't we just require that one? It is also mandatory in an HTTP response ...

Reply:
[Section 3.1](https://tools.ietf.org/html/rfc7807#section-3.1) of the RFC says:

>  The "status" member, if present, is only advisory; it conveys the
   HTTP status code used for the convenience of the consumer.
   Generators MUST use the same status code in the actual HTTP response,
   to assure that generic HTTP software that does not understand this
   format still behaves correctly.  See Section 5 for further caveats
   regarding its use.

Not having the Status as a option would reduce a lot of complexity. We can always provide a default for deserializing as well!

Reply:
That status method is "private" and gives an unused warning. I simply did not remove it because I was not sure. I also have the feeling that the distinction between the status as a number and the concept(the type) is not really consistent in the lib. That is why I did not remove the 2 private methods for now.

Regarding the optional status: Section 3.1 says that the status if it is there must be the same as in the response. I do not see a problem when we serialize since we always guarantee a valid status due to `http::StatusCode`. But what happens when we deserialize? There can be anything in the response. If we than just take 500 as a default the status in the JSON and the response might not match anymore which is a clear violation of "MUST use the same status.." on the receivers side. Unfortunately the http crate does not support "invalid/custom" status codes and fails instead.

I will merge the PR now and we can still change this later on. I would really appreciate your feedback. Next I plan to add another nasty feature toggle for actix. I there something to do for "warp"? 

FYI: I will not add that IntoStatusCode trait unless really needed. Mostly users generate the JSON and they should move to the HTTP crate.